### PR TITLE
WIP: Implementing the YIN algorithm

### LINKED ISSUE #92: YIN, a fundamental frequency estimator for speech and music
In #91 and #89 we implemented some methods to estimate the dominant period (or conversely dominant frequency) from a timeseries. @asinghvi17  suggested Yin as yet another method to do it, as it is described here:

http://audition.ens.fr/adc/pdf/2002_JASA_YIN.pdf
(implemented in Python at https://github.com/patriceguyot/Yin)

(for now I structure this as "low priority" only because we have already quite a bit of methods to find dominant period)

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/79487712-yin-a-fundamental-frequency-estimator-for-speech-and-music?utm_campaign=plugin&utm_content=tracker%2F81670622&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F81670622&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by anoojpatel**
Initial Commit with basic code structure. Please let me know if there is any styling requirements I need to adhere too, or if my structure needs to be rearranged/relocated!

Closes #92

### CODE REVIEW

**Comment by asinghvi17** [comment]
Most of this isn't actually working Julia code, there are a lot of `end`s missing for example.  Also, docstrings in Julia are formatted differently than those in Python.  

The structure we've had so far is that everything related to period estimation goes in the `period` folder, so you may want to move the file there.

Overall, if you can fix the syntax errors, it looks good for a first effort!

**Comment by FuturizeHandgun** [comment]
Thanks for the feedback! I wanted to get a WIP PR in as I get the code functional/tested and also cover needed style/organizational changes.

**Comment by asinghvi17** [comment]
I've made a PR on your repo, which fixes up some stuff.

**Comment by Datseris** [comment]
@asinghvi17 I am adding you as a reviewer just to make it clear that when you finish reviewing this, and you feel like it is done, you should go ahead and merge it without waiting for my input (I also look the PR and will make comments if need be).

Just be sure that there is a change in Project.toml and the CHANGELOG.md. :)

File: src/periodicity/yin.jl [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+"""
+    yin(
+        sig, sr;
+        w_len = 512,
+        w_step = 256,
+        f0_min = 100,
+        f0_max = 500,
+        harmonic_threshold = 0.1,
+        diffference_function = differenceFunction_original
```
Comment (Datseris) [inline src/periodicity/yin.jl]: There is no benefit of stating the keywords in the header. They have a dedicated section in the docstring.

Comment (Datseris) [reply src/periodicity/yin.jl]: You should use `kwargs...` instead.

**Comment by Datseris** [comment]
@FuturizeHandgun could you please restructure the docstring of `yin` to follow the same style of **DynamicalSystems.jl**:

> For the documentation strings of new methods and systems please follow the convention of the documentation strings of JuliaDynamics. Specifically, the first section should describe the function in a couple of sentences, its positional arguments and its return value. The next section `## Keyword Arguments` describes the keywords. The next section `## Description` describes the algorithm in detail if need be. A mandatory `## References` section lists the related literature that is cited.

as an example see: https://github.com/JuliaDynamics/ChaosTools.jl/blob/8602f67240aec15325a36a1f8a828f5f810ffe80/src/lyapunovs.jl#L9-L67

**Comment by Datseris** [comment]
@FuturizeHandgun bump. Would you like to move forward with this or should someone else take over?

**Comment by FuturizeHandgun** [comment]
@Datseris I would like to move forward with this! Sorry for my slow progress, I am looking for time in my schedule to contribute and learn how to also write test for this algorithm.

**Comment by Datseris** [comment]
Sure, take your time. Just ask if you need help!

**Comment by Datseris** [comment]
bump @anoojpatel  what is the status here?

**Comment by anoojpatel** [comment]
This has been low on my priority :( I don't mind if someone else has a further developed/working implementation who wants to take over. I don't think this will be completed by me soon.

**Comment by Datseris** [comment]
@asinghvi17 do you want to finish this? You should make your own PR if so.

**Pull request closed without merging**